src/lib/auth/AuthProviders/OAuthAuthenticator/LoginServer.ts

```diff
@@ -0,0 +1,185 @@
+    <meta charset="utf-8">
+  </head>
+  <body>
+    <p> Você já pode fechar essa janela. </p>
+    <p> You may now close this window. </p>
+    <p> Ahora puedes cerrar esta ventana. </p>
+  </body>
+</html>`
+
+export class LoginServer {
+  private static readonly LOGIN_CALLBACK_PATH = '/login_callback'
+  private static readonly SERVER_START_RETRIES = 1
+  private static readonly HOSTNAME = '127.0.0.1'
+
+  public static async create(loginConfig: LoginConfig) {
+    const loginServer = new LoginServer(loginConfig)
+    await loginServer.start()
+    return loginServer
+  }
+
+  private app: Koa
+  private port: number
+  private server: Server
+
+  private loginState?: string
+
+  private tokenPromise: Promise<string>
+  private resolveTokenPromise: (val: string) => void
+  private rejectTokenPromise: (err: any) => void
+
+  constructor(private loginConfig: LoginConfig) {
+    this.app = new Koa()
+    this.registerLoginHandler()
+
+    this.tokenPromise = new Promise((resolve, reject) => {
+      this.resolveTokenPromise = resolve
+      this.rejectTokenPromise = reject
+    })
+  }
+
+  get token() {
+    return this.tokenPromise
+  }
+
+  get loginCallbackUrl() {
+    if (!this.port) {
+      throw new Error('LoginServer not initialized')
+    }
+
+    return `http://${LoginServer.HOSTNAME}:${this.port}${LoginServer.LOGIN_CALLBACK_PATH}`
+  }
+
+  public setLoginState(val: string) {
+    this.loginState = val
+  }
+
+  public start() {
+    return asyncRetry(
+      async (bail, attemptNumber) => {
+        try {
+          // detectPort will get the specified port or, if it's in use, another ramdom unnused port
+          this.port = await detectPort(3000)
+          this.server = await this.initServer(this.port)
+          logger.debug(`LoginServer started on http://${LoginServer.HOSTNAME}:${this.port}`)
+        } catch (err) {
+          logger.debug(`LoginServer failed to start on port:${this.port}. Reason: ${err.message}.`)
+          if (err.code !== 'EADDRINUSE') {
+            return bail(err)
+          }
+
+          if (attemptNumber < LoginServer.SERVER_START_RETRIES + 1) {
+            logger.debug(`Retrying to start LoginServer...`)
+          }
+
+          throw ErrorReport.createAndMaybeRegisterOnTelemetry({
+            originalError: err,
+            kind: ErrorKinds.LOGIN_SERVER_START_ERROR,
+            details: { attemptNumber },
+          })
+        }
+      },
+      { retries: LoginServer.SERVER_START_RETRIES, maxTimeout: 100, minTimeout: 100 }
+    )
+  }
+
+  public close() {
+    this.server.unref()
+    this.server.close()
+  }
+
+  private initServer(port: number): Promise<Server> {
+    return new Promise((resolve, reject) => {
+      const server = this.app.listen(port, LoginServer.HOSTNAME, () => {
+        server.on('connection', socket => {
+          socket.unref()
+        })
+
+        server.removeListener('error', reject)
+        resolve(server)
+      })
+
+      server.on('error', reject)
```

Comment:
There was an error in the previous version - I was using `this.app.on('error', reject)`, and errors on server startup will be emitted by the server, not the app